osc_sdk/sdk.py

```diff
@@ -774,6 +816,78 @@ def build_headers(self, target, json_parameters):
         }
         return signed_headers, canonical_headers, headers
 
+
+class HttpsAdapter(requests.adapters.BaseAdapter):
+    """
+    Adapter to use an external SSL engine (thanks to M2Crypto)
+    https://docs.python-requests.org/en/master/_modules/requests/adapters/
+    """
+
+    def __init__(self, ssl_engine_id, ssl_engine_path, ssl_module_path):
+        super(HttpsAdapter, self).__init__()
+        self.ssl_engine_id = ssl_engine_id
+        self.ssl_engine_path = ssl_engine_path
+        self.ssl_module_path = ssl_module_path
+
+    # BaseAdapter method
+    def send(self, request, stream=False, timeout=None, verify=True, cert=None, proxies=None):
+
+        url_parse = urllib.parse.urlparse(request.url)
+        url_path_query = url_parse.path
+        if url_parse.query:
+            url_path_query += "?" + url_parse.query
+
+        cert_path = None
+        key_path = None
+        try:
+            (cert_path, key_path) = cert
+        except ValueError:
+            raise ValueError("Certificate and/or private key not configured")
+
+        ssl_context = self.init_ssl_context(verify=verify, cert_path=cert_path, key_path=key_path)
+        connection = M2Crypto.httpslib.HTTPSConnection(host=url_parse.netloc, ssl_context=ssl_context)
+        connection.request(method=request.method, url=url_path_query, body=request.body, headers=request.headers)
+
+        resp = urllib3.response.HTTPResponse.from_httplib(connection.getresponse())
+        response = requests.models.Response()
+        response.connection = self
+        response.url = url_parse.geturl()
+        response.headers = requests.structures.CaseInsensitiveDict(getattr(resp, "headers", {}))
+        response.reason = getattr(resp, "reason", None)
+        response.status_code = getattr(resp, "status", None)
+        response.request = request
+        response.raw = resp
+        response.encoding = requests.utils.get_encoding_from_headers(response.headers)
+        return response
+
+    # BaseAdapter method
+    def close(self):
+        pass
+
+    def init_ssl_context(self, verify, cert_path, key_path):
+        # More details on https://gitlab.com/m2crypto/m2crypto/-/blob/master/src/M2Crypto/SSL/Context.py
+        ssl_context = M2Crypto.SSL.Context()
+        if not verify:
+            ssl_context.set_allow_unknown_ca(True)
+
+        ssl_engine = self.init_ssl_engine()
+        cert = ssl_engine.load_certificate(cert_path)
+        key = ssl_engine.load_private_key(key_path)
+        M2Crypto.m2.ssl_ctx_use_x509(ssl_context.ctx, cert.x509) # pylint: disable=E1101 # is part of swig
+        M2Crypto.m2.ssl_ctx_use_pkey_privkey(ssl_context.ctx, key.pkey) # pylint: disable=E1101 # is part of swig
+        return ssl_context
+
+    def init_ssl_engine(self):
```

Comment:
is this function is used only once in the whole code, I'm not convince of it's need, maybe merge it inside init_ssl_context ?

Reply:
Would prefer to keep it as a separate function as it make its job clearly separated from `init_ssl_context`